File: cluster-fabric/src/main/java/dev/getelements/elements/cluster/fabric/JakartaWebsocketRemoteInvoker.java
package dev.getelements.elements.cluster.fabric;

import dev.getelements.elements.rt.PayloadReader;
import dev.getelements.elements.rt.PayloadWriter;
import dev.getelements.elements.rt.remote.*;
import jakarta.inject.Inject;
import jakarta.websocket.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

import static dev.getelements.elements.cluster.fabric.FabricInvocationFrame.Type.INVOCATION;

/**
 * The client side of the Fabric WebSocket transport (issue #10). Each invocation dials a fresh connection to the
 * peer's {@code /cluster/v1} endpoint, sends exactly one {@link Invocation} frame, receives exactly one response
 * frame, and closes &mdash; there is no persistent connection or connection pooling. Instances of this class are
 * minted per remote instance (one per {@link dev.getelements.elements.rt.remote.InstanceConnectionService.InstanceConnection}),
 * not shared globally, but a single instance is safe to use concurrently across many invocations since all
 * per-call state lives in a fresh {@link InvocationEndpoint} instance for each call.
 */
public class JakartaWebsocketRemoteInvoker implements RemoteInvoker {

    private static final Logger logger = LoggerFactory.getLogger(JakartaWebsocketRemoteInvoker.class);

    private static final ScheduledExecutorService timeoutScheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        final var thread = new Thread(r, JakartaWebsocketRemoteInvoker.class.getSimpleName() + " timeout scheduler");
        thread.setDaemon(true);
        return thread;
    });

    private final PayloadReader payloadReader;

    private final PayloadWriter payloadWriter;

    private volatile String connectAddress;

    private volatile long timeout = DEFAULT_TIMEOUT;

    private volatile TimeUnit timeoutTimeUnit = DEFAULT_TIMEOUT_UNITS;

    @Inject
    public JakartaWebsocketRemoteInvoker(final PayloadReader payloadReader, final PayloadWriter payloadWriter) {
        this.payloadReader = payloadReader;
        this.payloadWriter = payloadWriter;
    }

    @Override
    public String getConnectAddress() {
        return connectAddress;
    }

    @Override
    public void start(final String connectAddress, final long timeout, final TimeUnit timeoutTimeUnit) {
        this.connectAddress = connectAddress;
        this.timeout = timeout;
        this.timeoutTimeUnit = timeoutTimeUnit;
    }

    @Override
    public AsyncOperation invokeAsync(final Invocation invocation,
                                       final List<Consumer<InvocationResult>> asyncInvocationResultConsumerList,
                                       final InvocationErrorConsumer asyncInvocationErrorConsumer) {

        final var future = dial(invocation);

        future.whenComplete((invocationResult, throwable) -> {
            if (throwable == null) {
                asyncInvocationResultConsumerList.forEach(c -> c.accept(invocationResult));
            } else {
                asyncInvocationErrorConsumer.accept(toInvocationError(throwable));
            }
        });

        return new AsyncOperation() {

            @Override
            public void cancel() {
                future.cancel(true);
            }

            @Override
            public void timeout(final long time, final TimeUnit timeUnit) {
                timeoutScheduler.schedule(() -> future.cancel(true), time, timeUnit);
            }

        };

    }

    @Override
    public CompletionStage<Object> invokeCompletionStage(final Invocation invocation,
                                                          final List<Consumer<InvocationResult>> asyncInvocationResultConsumerList,
                                                          final InvocationErrorConsumer asyncInvocationErrorConsumer) {

        final var resultFuture = new CompletableFuture<Object>();

        dial(invocation).whenComplete((invocationResult, throwable) -> {
            if (throwable == null) {
                asyncInvocationResultConsumerList.forEach(c -> c.accept(invocationResult));
                resultFuture.complete(invocationResult.getResult());
            } else {
                asyncInvocationErrorConsumer.accept(toInvocationError(throwable));
                resultFuture.completeExceptionally(throwable);
            }
        });

        return resultFuture;

    }

    private static InvocationError toInvocationError(final Throwable throwable) {
        final var invocationError = new InvocationError();
        invocationError.setThrowable(throwable);
        return invocationError;
    }

    private CompletableFuture<InvocationResult> dial(final Invocation invocation) {

        final var endpoint = new InvocationEndpoint(payloadReader, payloadWriter, invocation);
        final var scheduledTimeout = timeoutScheduler.schedule(
                () -> endpoint.future.completeExceptionally(new java.util.concurrent.TimeoutException(
                        "Timed out after " + timeout + " " + timeoutTimeUnit + " connecting to " + connectAddress)),
                timeout, timeoutTimeUnit);

        endpoint.future.whenComplete((r, t) -> scheduledTimeout.cancel(false));

        try {
            final var container = ContainerProvider.getWebSocketContainer();
            container.connectToServer(endpoint, URI.create(connectAddress));
        } catch (final Exception ex) {
            endpoint.future.completeExceptionally(ex);
        }

        return endpoint.future;

    }

    @ClientEndpoint
    public static class InvocationEndpoint {

        private final PayloadReader payloadReader;

        private final PayloadWriter payloadWriter;

        private final Invocation invocation;

        private final CompletableFuture<InvocationResult> future = new CompletableFuture<>();

        public InvocationEndpoint(final PayloadReader payloadReader,
                                  final PayloadWriter payloadWriter,
                                  final Invocation invocation) {
            this.payloadReader = payloadReader;
            this.payloadWriter = payloadWriter;
            this.invocation = invocation;
        }

        @OnOpen
        public void onOpen(final Session session) {
            try {
                final var payload = payloadWriter.write(invocation);
                final var frame = new FabricInvocationFrame(INVOCATION, payload);
                session.getBasicRemote().sendBinary(ByteBuffer.wrap(frame.toBytes()));
            } catch (final IOException ex) {
                future.completeExceptionally(ex);
            }
        }

        @OnMessage
        public void onMessage(final Session session, final ByteBuffer message) {
            try {

                final var bytes = new byte[message.remaining()];
                message.get(bytes);
                final var frame = FabricInvocationFrame.fromBytes(bytes);

                switch (frame.getType()) {
                    case RESULT -> future.complete(payloadReader.read(InvocationResult.class, frame.getPayload()));
                    case ERROR -> {
                        final var error = payloadReader.read(InvocationError.class, frame.getPayload());
                        future.completeExceptionally(error.getThrowable());
                    }
                    default -> future.completeExceptionally(
                            new IllegalStateException("Unexpected Fabric frame type: " + frame.getType()));
                }

            } catch (final Exception ex) {
                future.completeExceptionally(ex);
            } finally {
                try {
                    session.close();
                } catch (final IOException ex) {
                    logger.debug("Failed to close session cleanly.", ex);
                }
            }
        }

        @OnError
        public void onError(final Session session, final Throwable thr) {
            future.completeExceptionally(thr);
        }

    }

}

File: cluster-client/src/main/kotlin/dev/getelements/elements/cluster/client/JakartaWebsocketRemoteInvoker.kt
package dev.getelements.elements.cluster.client

import dev.getelements.elements.cluster.common.dto.Envelope
import dev.getelements.elements.cluster.common.dto.Envelope.Type.INSTANCE_METADATA
import dev.getelements.elements.cluster.common.dto.Envelope.Type.INVOCATION_ERROR
import dev.getelements.elements.cluster.common.dto.Envelope.Type.INVOCATION_RESULT
import dev.getelements.elements.cluster.common.dto.InstanceMetadataEnvelope
import dev.getelements.elements.cluster.common.dto.InvocationErrorEnvelope
import dev.getelements.elements.cluster.common.dto.InvocationResultEnvelope
import dev.getelements.elements.sdk.Subscription
import dev.getelements.elements.sdk.cluster.remote.AsyncOperation
import dev.getelements.elements.sdk.cluster.remote.InstanceRemoteInvoker
import dev.getelements.elements.sdk.cluster.remote.InvocationErrorConsumer
import dev.getelements.elements.sdk.cluster.remote.RemoteInvoker
import dev.getelements.elements.sdk.cluster.remote.dto.InstanceMetadata
import dev.getelements.elements.sdk.cluster.remote.dto.Invocation
import dev.getelements.elements.sdk.cluster.remote.dto.InvocationResult
import dev.getelements.elements.sdk.util.ConcurrentDequePublisher
import jakarta.websocket.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.CompletionStage
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicReference
import java.util.function.BiConsumer
import java.util.function.Consumer

@ClientEndpoint
class JakartaWebsocketRemoteInvoker : InstanceRemoteInvoker {

    companion object {
        val logger: Logger = LoggerFactory.getLogger(JakartaWebsocketRemoteInvoker::class.java)
    }

    private val sequence = AtomicLong(0)

    private val session: AtomicReference<Session> = AtomicReference()

    private val operations : SortedMap<String, RemoteInvocationState> = TreeMap();

    private val instanceMetadata = AtomicReference(InstanceMetadata.DEFAULT)

    private val instanceMetadataPublisher = ConcurrentDequePublisher<InstanceMetadata>(JakartaWebsocketRemoteInvoker::class.java);

    private val closePublisher = ConcurrentDequePublisher<RemoteInvoker>(JakartaWebsocketRemoteInvoker::class.java)

    override fun invokeAsync(
        invocation: Invocation,
        asyncInvocationResultConsumerList: List<Consumer<InvocationResult>>,
        asyncInvocationErrorConsumer: InvocationErrorConsumer
    ): AsyncOperation {

        val id = sequence.getAndIncrement().toString()
        val session = this.session.get() ?: throw IllegalStateException("Session is null.")

        val operation = JakartaWebsocketAsyncOperation(
            id,
            session,
            invocation,
            asyncInvocationResultConsumerList,
            asyncInvocationErrorConsumer
        )

        operations[operation.id] = operation
        operation.send()

        return operation

    }

    override fun invokeCompletionStage(
        invocation: Invocation,
        asyncInvocationResultConsumerList: List<Consumer<InvocationResult>>,
        asyncInvocationErrorConsumer: InvocationErrorConsumer
    ): CompletionStage<Any> {

        val id = sequence.getAndIncrement().toString()
        val session = this.session.get() ?: throw IllegalStateException("Session is null.")

        val invocation = JakartaWebsocketRemoteInvocation(
            id,
            session,
            invocation,
            asyncInvocationResultConsumerList,
            asyncInvocationErrorConsumer
        )

        operations[invocation.id] = invocation
        return invocation.send()

    }

    @OnOpen
    suspend fun onOpen(session: Session) {
        this.session.set(session)
    }

    @OnError
    suspend fun onError(session: Session, throwable: Throwable) {
        operations.values.forEach { it.onError(throwable) }
    }

    @OnClose
    suspend fun onClose(session: Session) {
        closePublisher.publish(this)
    }

    @OnMessage
    suspend fun onMessage(session: Session, envelope: Envelope<Any>) {
        when (envelope.type) {
            INVOCATION_ERROR  -> onInvocationError(envelope as InvocationErrorEnvelope)
            INVOCATION_RESULT -> onInvocationResult(envelope as InvocationResultEnvelope)
            INSTANCE_METADATA -> onInstanceMetadata(envelope as InstanceMetadataEnvelope)
            else -> throw IllegalArgumentException("Envelope type ${envelope.type} is not supported")
        }
    }

    private suspend fun onInvocationError(envelope: InvocationErrorEnvelope) {
        val operation = operations[envelope.id]
        if (operation == null) logger.error("Unknown result invocation {}", envelope.id)
        operation?.onInvocationError(envelope)
    }

    private suspend fun onInvocationResult(envelope: InvocationResultEnvelope) {
        val operation = operations[envelope.id]
        if (operation == null) logger.error("Unknown result invocation {}", envelope.id)
        operation?.onInvocationResult(envelope)
    }

    private fun onInstanceMetadata(envelope: InstanceMetadataEnvelope) {
        instanceMetadata.set(envelope.payload)
        instanceMetadataPublisher.publish(envelope.payload)
    }

    private fun getInvocationState(id:String) : RemoteInvocationState {
        return operations[id] ?: throw IllegalStateException("Session is null.")
    }

    override fun getInstanceMetadata(): InstanceMetadata? =
        instanceMetadata.get()

    override fun onMetadataUpdate(onMetadata: Consumer<InstanceMetadata?>?): Subscription? =
        instanceMetadataPublisher.subscribe(onMetadata)

    override fun onMetadataUpdate(onMetadata: BiConsumer<Subscription?, InstanceMetadata?>?): Subscription? =
        instanceMetadataPublisher.subscribe(onMetadata)

    override fun onClose(remoteInvokerConsumer: Consumer<RemoteInvoker>): Subscription =
        closePublisher.subscribe(remoteInvokerConsumer)

    override fun onClose(remoteInvokerBiConsumer: BiConsumer<Subscription, RemoteInvoker>): Subscription =
        closePublisher.subscribe(remoteInvokerBiConsumer)

}
